### DISCUSSION TITLE
How to mock external authentication decorator used along with routes

### DISCUSSION THREAD

**Discussion by David-Vithayathil** [fastapi/fastapi#11584] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# project_name/routes/school_routes.py
from external_library import authentication

@router.get("/", status_code=status.HTTP_200_OK, tags=["schools"])
@authentication
def list_schools(request: Request):
  #Code logic

# tests/routes.py
client = TestClient(app)

class TestRoutes(unittest.TestCase):
  @patch('project_name.routes.school_routes.authentication')
  def setUp(self, mock_auth):
     mock_auth.return_value = lambda func: func

  def test_list_schools(self):
     response = client.get(<PATH>)
```

### Description

I am importing this authentication decorator from an external library. I am currently writing unittests for the application and somehow the authentication decorator doesn't get patched.

I am aware of using authentication through Depends() but somehow I am not permitted to alter the authentication decorator as it is commonly used in several other projects primarily written using other frameworks within the org.

### Operating System

macOS

### FastAPI Version

0.110.3

### Pydantic Version

2.7.1

### Python Version

3.10

**Comment by aanchlia** [comment]
Have you tried using `patch`? Something like this below.

```python
import unittest
from unittest.mock import patch

patcher = patch('project_name.routes.school_routes.authentication')
self.mock_auth = patcher.start()


```

**Reply by David-Vithayathil** [reply]
Yes. I've used the patch decorator. The call is still going to the original authentication decorator and not the patched one.

**Comment by YuriiMotov** [comment, marked as answer]
You can't patch decorator since Python applies decorators right after you import the module and before methods of `TestRoutes` are called.
So, at the moment when you call `TestRoutes`'s methods, `list_schools` is already modified and patching the `authentication` function doesn't have any effect.

Your case is a clear example of the benefits of using dependencies instead of other mechanisms (injecting logic via middlewares, decorators, etc..). 

I think you can overcome this by introducing a dependency like it's shown above:
```
# project_name/routes/school_routes.py
from external_library import authentication

@authentication
def authenticate(request: Request):
    pass

@router.get("/", status_code=status.HTTP_200_OK, tags=["schools"], dependencies=[Depends(authenticate)])
def list_schools():
  #Code logic
```

To mock the authentication just override the dependency:
```
def mock_authenticate():
    pass

app.dependency_overrides[authenticate] = mock_authenticate
```

**Reply by David-Vithayathil** [reply]
Thanks a lot. I've tried implementing it as you mentioned and it works !!
